src/cpp/session/modules/build/SessionBuild.cpp

```diff
@@ -1012,6 +1046,110 @@ class Build : boost::noncopyable,
 
    }
 
+   void testFile(const FilePath& testPath,
+                 core::system::ProcessOptions pkgOptions,
+                 const core::system::ProcessCallbacks& cb)
+   {
+      s_buildEnabled = true;
+      ClientEvent event(client_events::kEnableBuild);
+      module_context::enqueClientEvent(event);
+
+      FilePath rScriptPath;
+      Error error = module_context::rScriptPath(&rScriptPath);
+      if (error)
+      {
+         terminateWithError("Locating R script", error);
+         return;
+      }
+
+      // construct a shell command to execute
+      shell_utils::ShellCommand cmd(rScriptPath);
+      cmd << "--slave";
+      cmd << "--vanilla";
+      cmd << "-e";
+      std::vector<std::string> rSourceCommands;
+      
+      boost::format fmt(
+         "if (nchar('%1%')) library('%1%');"
+         "message('Running: testthat::test_file(%2%)');"
+         "testthat::test_file('%2%')"
+      );
+
+      cmd << boost::str(fmt %
+                        pkgInfo_.name() %
+                        testPath.absolutePath());
```

Comment:
The path needs to be converted into the system encoding, and have single quotes escaped.